app/src/main/kotlin/com/wire/android/ui/authentication/welcome/Welcome.kt

```diff
@@ -98,32 +101,41 @@ private fun WelcomeContent(navController: NavController) {
     }
 }
 
-@OptIn(ExperimentalPagerApi::class, ExperimentalFoundationApi::class)
+@OptIn(ExperimentalPagerApi::class, ExperimentalFoundationApi::class, ExperimentalCoroutinesApi::class)
 @Composable
 private fun WelcomeCarousel() {
     val delay = integerResource(id = R.integer.welcome_carousel_item_time_ms)
     val icons: List<Int> = typedArrayResource(id = R.array.welcome_carousel_icons).drawableResIdList()
     val texts: List<String> = stringArrayResource(id = R.array.welcome_carousel_texts).toList()
     val items: List<Pair<Int, String>> = icons zip texts
-    val circularItemsList = listOf<Pair<Int, String>>().plus(items.last()).plus(items)
-    val pageState = rememberPagerState(initialPage = 1)
-
-    LaunchedEffect(pageState.currentPage) {
-        if (pageState.currentPage == circularItemsList.lastIndex) {
-            pageState.scrollToPage(0)
-        } else {
-            delay(delay.toLong())
-            pageState.animateScrollToPage(pageState.currentPage + 1)
-        }
+    val circularItemsList = listOf<Pair<Int, String>>().plus(items.last()).plus(items).plus(items.first())
+    val initialPage = 1
+    val pageState = rememberPagerState(initialPage = initialPage)
+
+    LaunchedEffect(pageState) {
+        snapshotFlow { pageState.currentPage }
+            .distinctUntilChanged()
+            .scan(initialPage to initialPage) { (_, lastPage), currentPage -> lastPage to currentPage }
+            .flatMapLatest { (lastPage, currentPage) ->
+                if (currentPage == circularItemsList.lastIndex && lastPage < currentPage && lastPage >= initialPage)
```

Comment:
can you extract this `if` into some `fun` with meaning name, like `if(shouldJumpToStart())` or something like that